File: sdk/samples/02_qos/python/transport_priority.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0 OR MIT
# Copyright (c) 2025-2026 naskel.com

"""
HDDS Sample: Transport Priority (Python)

Demonstrates TRANSPORT_PRIORITY QoS for network priority levels.
Alarm data uses high priority (10), telemetry uses low priority (0).
Priority mapping depends on OS support (e.g., DSCP/TOS bits).

Usage:
    python transport_priority.py        # Subscriber (monitors both topics)
    python transport_priority.py pub    # Publisher (sends on both topics)
"""

import sys
import time

sys.path.insert(0, str(__file__ + "/../../01_basics/python"))

import hdds
from generated.HelloWorld import HelloWorld

PRIORITY_HIGH = 10  # Alarm data
PRIORITY_LOW = 0    # Telemetry data
NUM_MESSAGES = 5


def run_publisher(participant):
    """Publish alarm (high priority) and telemetry (low priority) data."""
    qos_alarm = hdds.QoS.reliable().transport_priority(PRIORITY_HIGH)
    qos_telemetry = hdds.QoS.reliable().transport_priority(PRIORITY_LOW)

    writer_alarm = participant.create_writer("AlarmTopic", qos=qos_alarm)
    writer_telemetry = participant.create_writer("TelemetryTopic", qos=qos_telemetry)

    print(f"Publishing {NUM_MESSAGES} messages on each topic...")
    print(f"  AlarmTopic:     priority = {PRIORITY_HIGH} (high)")
    print(f"  TelemetryTopic: priority = {PRIORITY_LOW} (low)\n")

    start = time.monotonic()

    # Send interleaved bursts: telemetry first, then alarm
    for i in range(NUM_MESSAGES):
        elapsed = int((time.monotonic() - start) * 1000)

        msg_tel = HelloWorld(id=i + 1, message=f"Telemetry #{i + 1}")
        writer_telemetry.write(msg_tel.serialize())
        print(f"  [{elapsed:5d}ms] Sent Telemetry (priority={PRIORITY_LOW})  id={msg_tel.id}")

        msg_alarm = HelloWorld(id=i + 1, message=f"ALARM #{i + 1}")
        writer_alarm.write(msg_alarm.serialize())
        print(f"  [{elapsed:5d}ms] Sent Alarm     (priority={PRIORITY_HIGH}) id={msg_alarm.id}")

        time.sleep(0.2)

    print("\nDone publishing.")
    print("Note: With OS/DSCP support, alarm data may arrive before telemetry.")


def run_subscriber(participant):
    """Monitor both topics and observe arrival order."""
    qos_alarm = hdds.QoS.reliable().transport_priority(PRIORITY_HIGH)
    qos_telemetry = hdds.QoS.reliable().transport_priority(PRIORITY_LOW)

    reader_alarm = participant.create_reader("AlarmTopic", qos=qos_alarm)
    reader_telemetry = participant.create_reader("TelemetryTopic", qos=qos_telemetry)

    waitset = hdds.WaitSet()
    waitset.attach(reader_alarm)
    waitset.attach(reader_telemetry)

    print("Subscribing to both topics...")
    print(f"  AlarmTopic:     priority = {PRIORITY_HIGH} (high)")
    print(f"  TelemetryTopic: priority = {PRIORITY_LOW} (low)")
    print("\nMonitoring arrival order...\n")

    received_alarm = 0
    received_telemetry = 0
    arrival_order = []
    start = time.monotonic()
    timeouts = 0

    while timeouts < 3:
        if waitset.wait(timeout_secs=2.0):
            while True:
                data = reader_alarm.take()
                if data is None:
                    break
                msg = HelloWorld.deserialize(data)
                elapsed = int((time.monotonic() - start) * 1000)
                print(f"  [{elapsed:5d}ms] ALARM     received id={msg.id}")
                arrival_order.append(f"ALARM-{msg.id}")
                received_alarm += 1

            while True:
                data = reader_telemetry.take()
                if data is None:
                    break
                msg = HelloWorld.deserialize(data)
                elapsed = int((time.monotonic() - start) * 1000)
                print(f"  [{elapsed:5d}ms] Telemetry received id={msg.id}")
                arrival_order.append(f"TEL-{msg.id}")
                received_telemetry += 1

            timeouts = 0
        else:
            timeouts += 1

    print()
    print("-" * 60)
    print(f"Summary:")
    print(f"  Alarm     (priority={PRIORITY_HIGH}): {received_alarm} messages")
    print(f"  Telemetry (priority={PRIORITY_LOW}):  {received_telemetry} messages")
    print(f"  Arrival order: {' -> '.join(arrival_order[:10])}")
    print()
    print("Note: Transport priority maps to OS-level DSCP/TOS bits.")
    print("Actual prioritization depends on network stack and QoS support.")
    print("Without OS support, arrival order may appear identical.")
    print("-" * 60)


def main():
    is_publisher = len(sys.argv) > 1 and sys.argv[1] == "pub"

    hdds.logging.init(hdds.LogLevel.INFO)

    print("=" * 60)
    print("Transport Priority Demo")
    print("QoS: TRANSPORT_PRIORITY - network-level priority hints")
    print("=" * 60)

    with hdds.Participant("TransportPriorityDemo") as participant:
        if is_publisher:
            run_publisher(participant)
        else:
            run_subscriber(participant)


if __name__ == "__main__":
    main()

File: sdk/samples/02_qos/rust/src/bin/transport_priority.rs
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright (c) 2025-2026 naskel.com

//! # HDDS Sample: Transport Priority QoS
//!
//! Demonstrates **TRANSPORT_PRIORITY** QoS - assigns network priority levels
//! to data flows, enabling quality-of-service differentiation at the
//! transport layer.
//!
//! ## How TRANSPORT_PRIORITY Works
//!
//! ```text
//!                     Priority Queues
//!                   +------------------+
//! Alarm (pri=10) -->| HIGH  [A][A][A]  |---> Network (DSCP EF)
//!                   +------------------+
//!                   |                  |
//! Telemetry (0)  -->| LOW   [T][T][T]  |---> Network (DSCP BE)
//!                   +------------------+
//!
//! Under congestion, high-priority traffic is sent first.
//! OS and NIC must support traffic classification for full effect.
//! ```
//!
//! ## Use Cases
//!
//! - **Emergency data**: Alarms and safety signals get priority
//! - **DSCP marking**: Map DDS priority to IP DiffServ code points
//! - **Traffic class separation**: Isolate critical from bulk traffic
//! - **QoS-aware networks**: Leverage network-level prioritization
//!
//! ## Running the Sample
//!
//! ```bash
//! # Single-process mode (recommended):
//! cargo run --bin transport_priority
//!
//! # Two-terminal mode:
//! # Terminal 1 - Subscriber
//! cargo run --bin transport_priority -- sub
//!
//! # Terminal 2 - Publisher
//! cargo run --bin transport_priority -- pub
//! ```

use std::env;
use std::sync::Arc;
use std::thread;
use std::time::{Duration, Instant};

#[allow(dead_code)]
mod generated {
    include!("../../../../01_basics/rust/generated/hello_world.rs");
}

use generated::hdds_samples::HelloWorld;

/// Number of messages per priority level
const NUM_MESSAGES: u32 = 5;
/// High priority value (maps to DSCP Expedited Forwarding)
const PRIORITY_HIGH: i32 = 10;
/// Low priority value (maps to DSCP Best Effort)
const PRIORITY_LOW: i32 = 0;

// =============================================================================
// Publisher
// =============================================================================

fn run_publisher(participant: &Arc<hdds::Participant>) -> Result<(), hdds::Error> {
    // -------------------------------------------------------------------------
    // High-Priority Writer (Alarms)
    // -------------------------------------------------------------------------
    //
    // Transport priority = 10. On supported networks, this maps to higher
    // DSCP values, giving these packets preferential treatment.

    let alarm_qos = hdds::QoS::reliable().transport_priority(PRIORITY_HIGH);

    let alarm_writer = participant.create_writer::<HelloWorld>("AlarmTopic", alarm_qos)?;

    // -------------------------------------------------------------------------
    // Low-Priority Writer (Telemetry)
    // -------------------------------------------------------------------------
    //
    // Transport priority = 0. Standard best-effort network treatment.

    let telemetry_qos = hdds::QoS::reliable().transport_priority(PRIORITY_LOW);

    let telemetry_writer =
        participant.create_writer::<HelloWorld>("TelemetryTopic", telemetry_qos)?;

    println!(
        "Publishing {} messages on each priority level...\n",
        NUM_MESSAGES
    );

    let start = Instant::now();

    // Send a burst of interleaved high and low priority messages
    for i in 0..NUM_MESSAGES {
        let elapsed = start.elapsed().as_millis();

        // High-priority alarm
        let alarm_msg = HelloWorld::new(format!("ALARM #{}", i + 1), i + 1);
        alarm_writer.write(&alarm_msg)?;
        println!(
            "  [{:5}ms] Sent ALARM     #{} (priority={})",
            elapsed,
            i + 1,
            PRIORITY_HIGH
        );

        // Low-priority telemetry
        let telem_msg = HelloWorld::new(format!("Telemetry #{}", i + 1), i + 1);
        telemetry_writer.write(&telem_msg)?;
        println!(
            "  [{:5}ms] Sent TELEMETRY #{} (priority={})",
            elapsed,
            i + 1,
            PRIORITY_LOW
        );

        thread::sleep(Duration::from_millis(100));
    }

    println!("\nPublisher finished.");
    Ok(())
}

// =============================================================================
// Subscriber
// =============================================================================

fn run_subscriber(participant: &Arc<hdds::Participant>) -> Result<(), hdds::Error> {
    // -------------------------------------------------------------------------
    // Alarm Reader (high priority)
    // -------------------------------------------------------------------------

    let alarm_qos = hdds::QoS::reliable().transport_priority(PRIORITY_HIGH);

    let alarm_reader = participant.create_reader::<HelloWorld>("AlarmTopic", alarm_qos)?;

    // -------------------------------------------------------------------------
    // Telemetry Reader (low priority)
    // -------------------------------------------------------------------------

    let telemetry_qos = hdds::QoS::reliable().transport_priority(PRIORITY_LOW);

    let telemetry_reader =
        participant.create_reader::<HelloWorld>("TelemetryTopic", telemetry_qos)?;

    let waitset = hdds::dds::WaitSet::new();
    waitset.attach_condition(alarm_reader.get_status_condition())?;
    waitset.attach_condition(telemetry_reader.get_status_condition())?;

    println!("Listening for alarm and telemetry data...\n");

    let mut alarm_count = 0u32;
    let mut telemetry_count = 0u32;
    let mut arrival_order: Vec<String> = Vec::new();
    let start = Instant::now();
    let mut timeouts = 0;

    while (alarm_count + telemetry_count) < NUM_MESSAGES * 2 && timeouts < 3 {
        match waitset.wait(Some(Duration::from_secs(2))) {
            Ok(triggered) if !triggered.is_empty() => {
                let elapsed = start.elapsed().as_millis();

                // Check alarms first (higher priority)
                while let Some(msg) = alarm_reader.take().ok().flatten() {
                    alarm_count += 1;
                    arrival_order.push(format!("ALARM#{}", msg.count));
                    println!(
                        "  [{:5}ms] Received ALARM     #{} (priority={})",
                        elapsed, msg.count, PRIORITY_HIGH
                    );
                }

                // Check telemetry
                while let Some(msg) = telemetry_reader.take().ok().flatten() {
                    telemetry_count += 1;
                    arrival_order.push(format!("TELEM#{}", msg.count));
                    println!(
                        "  [{:5}ms] Received TELEMETRY #{} (priority={})",
                        elapsed, msg.count, PRIORITY_LOW
                    );
                }

                timeouts = 0;
            }
            Ok(_) | Err(hdds::Error::WouldBlock) => {
                timeouts += 1;
            }
            Err(e) => eprintln!("Wait error: {:?}", e),
        }
    }

    // -------------------------------------------------------------------------
    // Summary
    // -------------------------------------------------------------------------

    println!("\n{}", "-".repeat(50));
    println!(
        "Alarms received    (priority={}): {}",
        PRIORITY_HIGH, alarm_count
    );
    println!(
        "Telemetry received (priority={}):  {}",
        PRIORITY_LOW, telemetry_count
    );
    println!("{}", "-".repeat(50));
    println!("Arrival order: {:?}", arrival_order);
    println!("{}", "-".repeat(50));
    println!("Note: On a congested network with DSCP-aware switches,");
    println!("high-priority alarms would arrive before telemetry.");
    println!("In IntraProcess mode, ordering depends on write order.");
    println!("{}", "-".repeat(50));

    Ok(())
}

// =============================================================================
// Single-Process Demo
// =============================================================================

fn run_single_process(participant: &Arc<hdds::Participant>) -> Result<(), hdds::Error> {
    // -------------------------------------------------------------------------
    // Setup: two writers (alarm + telemetry), two readers
    // -------------------------------------------------------------------------

    let alarm_qos = hdds::QoS::reliable().transport_priority(PRIORITY_HIGH);

    let telemetry_qos = hdds::QoS::reliable().transport_priority(PRIORITY_LOW);

    let alarm_writer = participant.create_writer::<HelloWorld>("AlarmTopic", alarm_qos.clone())?;
    let telemetry_writer =
        participant.create_writer::<HelloWorld>("TelemetryTopic", telemetry_qos.clone())?;

    let alarm_reader = participant.create_reader::<HelloWorld>("AlarmTopic", alarm_qos)?;
    let telemetry_reader =
        participant.create_reader::<HelloWorld>("TelemetryTopic", telemetry_qos)?;

    let waitset = hdds::dds::WaitSet::new();
    waitset.attach_condition(alarm_reader.get_status_condition())?;
    waitset.attach_condition(telemetry_reader.get_status_condition())?;

    // Allow time for discovery
    thread::sleep(Duration::from_millis(100));

    // -------------------------------------------------------------------------
    // Send burst of interleaved messages
    // -------------------------------------------------------------------------

    println!(
        "Sending burst of {} messages per priority level...\n",
        NUM_MESSAGES
    );

    let start = Instant::now();

    for i in 0..NUM_MESSAGES {
        let elapsed = start.elapsed().as_millis();

        // Interleave: telemetry first, then alarm (to see if priority reorders)
        let telem_msg = HelloWorld::new(format!("Telemetry #{}", i + 1), i + 1);
        telemetry_writer.write(&telem_msg)?;

        let alarm_msg = HelloWorld::new(format!("ALARM #{}", i + 1), i + 1);
        alarm_writer.write(&alarm_msg)?;

        println!(
            "  [{:5}ms] Sent TELEMETRY #{} then ALARM #{}",
            elapsed,
            i + 1,
            i + 1
        );
    }

    // Read back and track arrival order
    thread::sleep(Duration::from_millis(100));

    let mut alarm_count = 0u32;
    let mut telemetry_count = 0u32;
    let mut arrival_order: Vec<String> = Vec::new();
    let mut timeouts = 0;

    while (alarm_count + telemetry_count) < NUM_MESSAGES * 2 && timeouts < 3 {
        match waitset.wait(Some(Duration::from_secs(2))) {
            Ok(triggered) if !triggered.is_empty() => {
                let elapsed = start.elapsed().as_millis();

                while let Some(msg) = alarm_reader.take().ok().flatten() {
                    alarm_count += 1;
                    arrival_order.push(format!("ALARM#{}", msg.count));
                    println!("  [{:5}ms] Received ALARM     #{}", elapsed, msg.count);
                }

                while let Some(msg) = telemetry_reader.take().ok().flatten() {
                    telemetry_count += 1;
                    arrival_order.push(format!("TELEM#{}", msg.count));
                    println!("  [{:5}ms] Received TELEMETRY #{}", elapsed, msg.count);
                }

                timeouts = 0;
            }
            Ok(_) | Err(hdds::Error::WouldBlock) => {
                timeouts += 1;
            }
            Err(e) => eprintln!("Wait error: {:?}", e),
        }
    }

    // -------------------------------------------------------------------------
    // Summary
    // -------------------------------------------------------------------------

    println!("\n{}", "-".repeat(50));
    println!(
        "Alarms received    (priority={}): {}",
        PRIORITY_HIGH, alarm_count
    );
    println!(
        "Telemetry received (priority={}):  {}",
        PRIORITY_LOW, telemetry_count
    );
    println!("{}", "-".repeat(50));
    println!("Arrival order: {:?}", arrival_order);
    println!("{}", "-".repeat(50));
    println!("Note: Transport priority affects DSCP/TOS bits in IP packets.");
    println!("For actual prioritization, configure your OS and network:");
    println!("  - Linux: tc qdisc, SO_PRIORITY socket option");
    println!("  - Switches: DSCP-to-queue mapping");
    println!("In IntraProcess mode, no network prioritization occurs.");
    println!("{}", "-".repeat(50));

    Ok(())
}

// =============================================================================
// Main
// =============================================================================

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    let mode = args.get(1).map(|s| s.as_str());

    println!("{}", "=".repeat(60));
    println!("HDDS Transport Priority QoS Sample");
    println!("Network priority levels for traffic differentiation");
    println!("{}\n", "=".repeat(60));

    let participant = hdds::Participant::builder("TransportPriorityDemo")
        .with_transport(hdds::TransportMode::IntraProcess)
        .build()?;

    match mode {
        Some("pub") => run_publisher(&participant)?,
        Some("sub") => run_subscriber(&participant)?,
        _ => run_single_process(&participant)?,
    }

    Ok(())
}
